streamz/core.py

```diff
@@ -1131,19 +1272,26 @@ def _remove_upstream(self, upstream):
                                "``node.emit_on=tuple(node.upstreams)`` to "
                                "emit on all incoming data")
         self.last.pop(self.upstreams.index(upstream))
+        self.metadata.pop(self.upstreams.index(upstream))
         self.missing.remove(upstream)
         super(combine_latest, self)._remove_upstream(upstream)
         if self._initial_emit_on is None:
             self.emit_on = self.upstreams
 
-    def update(self, x, who=None):
+    def update(self, x, who=None, metadata=None):
+        self._retain_refs(metadata)
+        idx = self.upstreams.index(who)
+        if self.metadata[idx]:
+            self._release_refs(self.metadata[idx])
```

Comment:
Is this releasing the reference too early?
It seems that you would need the reference until the tuple is emitted (line 1294)

Reply:
I think we would need to release the reference when we see a new element in the stream that would now be the 'latest' for that stream. So this the current code is correct?

Reply:
Ah ok, sorry I misread that. I would move this down so that it is matched up with `self.last[idx] = x` which is doing the same thing but for the data. Or maybe this needs a comment saying that we're bouncing an element out from that position (if there is an element at that position) and therefore need to release the reference count

Reply:
I do not think this is too early. Here `self.metadata` holds the metadata for the previous call to `update`. When a new value comes in, that new value is emitted and not the value that is in the `self.metadata` buffer. The new data replaces the value in the buffer and the old value is "released" as it is no longer being "retained" by the `combine_latest` node.